### PR TITLE
Nodes for Donatello

### PR DESCRIPTION
**Pull request by i-make-robots**
[Donatello](https://github.com/MarginallyClever/NodeGraphCore) has a service to register nodes and data types.
This branch adds nodes, tests for those nodes, sets up the META-INF/services, and the module-info.java needed to export the service.
The finished makelangelo-software-*.jar can be copied to `~/Donatello/extensions/` and then Donatello will find and load the nodes like any other plugin.

### CODE REVIEW

**Comment by i-make-robots** [comment]
`com.marginallyClever.makelangelo.paper.PaperSettingsTest.runPaperSettings()` is failing with headless exception.  I don't remember how to disable some tests when headless.

`com.marginallyClever.makelangelo.plotter.plotterControls.SaveGCodeTest` are all failing.

File: src/main/java/com/marginallyclever/donatello/nodes/PathImageMask.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+     * @param img the image to sample while converting along the line.
+     * @param segment the line to walk.
+     * @param stepSize millimeters level of detail for this line.
+     * @param channelCutoff only put pen down when color below this amount.
+     */
+    private void scanLine(BufferedImage img, LineSegment2D segment, double stepSize, double channelCutoff) {
+        Point2D P0 = segment.start;
+        Point2D P1 = segment.end;
+
+        ArrayList<LineSegment2D> toKeep = new ArrayList<>();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/donatello/nodes/PathImageMask.java:90]:
```suggestion
        List<LineSegment2D> toKeep = new ArrayList<>();
```

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public LoadTurtle(String filename) {
+        this();
+        this.filename.setValue(filename);
+    }
+
+    @Override
+    public Node create() {
+        return new LoadTurtle();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java:31]: this method is constructed in the same way in each class, have you tried to transform `Node` to a generic type and let this class instantiate it like :  `public class LoadTurtle extends Node<LoadTurtle>  {`

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public void update() {
+        try {
+            Turtle t = TurtleFactory.load(filename.getValue());
+            contents.setValue(t);
+            Rectangle2D r = t.getBounds();
+            w.setValue(r.getWidth());
+            h.setValue(r.getHeight());
+            cleanAllInputs();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java:44]: please use the logger.

File: src/main/java/com/marginallyclever/donatello/nodes/PatternOnPath.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    n+=step;
+                    Turtle stamp = new Turtle(myPattern);
+                    stamp.translate(p.x,p.y);
+                    sum.add(stamp);
+                }
+            }
+            output.setValue(sum);
+            cleanAllInputs();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/donatello/nodes/PatternOnPath.java:61]: please use the logger. I will add a remark for the following `catch`. If you don't want to add a `try catch` in each implementation, move it to the caller instead

File: src/main/java/com/marginallyclever/donatello/nodes/TurtleJSON_DAO.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.marginallyClever.donatello.nodes;
+
+import com.marginallyClever.nodeGraphCore.JSON_DAO;
+import com.marginallyClever.makelangelo.turtle.Turtle;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class TurtleJSON_DAO implements JSON_DAO<Turtle> {
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/donatello/nodes/TurtleJSON_DAO.java:8]: the name of a class must not contains `-` or `_`

File: src/main/java/com/marginallyclever/makelangelo/makeArt/InfillTurtle.java [hunk 6]
```diff
@@ -105,25 +105,25 @@ private void confirmTurtleIsClosedLoop(Turtle input) throws Exception {
...
+	private ArrayList<LineSegment2D> trimLineToPath(LineSegment2D line, List<LineSegment2D> convertedPath) {
+		ArrayList<Point2D> intersections = new ArrayList<>();
 
 		for (LineSegment2D s : convertedPath) {
 			Point2D p = getIntersection(line, s);
 			if (p != null) intersections.add(p);
 		}
 
-		ArrayList<LineSegment2D> results = new ArrayList<LineSegment2D>();
+		ArrayList<LineSegment2D> results = new ArrayList<>();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/makeArt/InfillTurtle.java:120]:
```suggestion
		List<LineSegment2D> results = new ArrayList<>();
```

File: src/main/java/module-info.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+module com.marginallyClever.makelangelo {
```
Comment (coliss86) [inline src/main/java/module-info.java:1]: I thought we can get ride of this file :-(

**Review by coliss86 — CHANGES_REQUESTED** [review]
To much changed files, hard to read. You should have split in 2 : one for the renaming of the package, and one for this new feature.
I see only one or two unit tests for all those new classes, you also should write more unit tests.
The CI is broken, please fix it.

Please add a documentation to explain how to use it, I don't understand. maybe you have planned to do it in another PR.

Good job !

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java (hunk 3 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java:44]: I have not yet resolved how to log from here if the code is run from another donatello, which does not yet use the logger.

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java (hunk 2 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java:31]: how would the addVariable() get called?

File: src/main/java/com/marginallyclever/donatello/nodes/PathImageMask.java (hunk 1 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/donatello/nodes/PathImageMask.java:90]: done.

File: src/main/java/com/marginallyclever/donatello/nodes/PatternOnPath.java (hunk 4 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/donatello/nodes/PatternOnPath.java:61]: Moved to caller.

File: src/main/java/com/marginallyclever/donatello/nodes/TurtleJSON_DAO.java (hunk 5 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/donatello/nodes/TurtleJSON_DAO.java:8]: changed all *_JSON_DAO and related to *DAO4JSON

**Comment by i-make-robots** [comment]
> The CI is broken, please fix it.

Unfortunately I didn't change the stuff related to the tests in my work.  I don't know why they're broken.  I am looking but I will be slower than the last person who altered these methods.

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java (hunk 2 above)
Comment (coliss86) [reply src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java:31]: where do you have this ?

File: src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java (hunk 3 above)
Comment (coliss86) [reply src/main/java/com/marginallyclever/donatello/nodes/LoadTurtle.java]: the other app should use a logger too :-)

**Comment by coliss86** [comment]
> > The CI is broken, please fix it.
> 
> Unfortunately I didn't change the stuff related to the tests in my work. I don't know why they're broken. I am looking but I will be slower than the last person who altered these methods.

Sure you will be slower but you will learn how it works and maybe it will give you some idea for futures tests

**Comment by i-make-robots** [comment]
https://github.community/t/maven-workflow-unauthorized-to-download-dependency-from-github-packages/238486

**Pull request merged**